File: Lab01/ejercicios_resueltos/Anexo2/Python/cubbyhole.py
import threading

class CubbyHole:
    def __init__(self):
        self.contents = 0
        self.available = False
        self.condition = threading.Condition()

    def get(self):
        with self.condition:
            while not self.available:
                self.condition.wait()  # Espera a que el productor ponga algo
            
            self.available = False
            self.condition.notify_all()  # Avisa al productor
            return self.contents

    def put(self, value):
        with self.condition:
            while self.available:
                self.condition.wait()  # Espera a que el consumidor saque el dato
            
            self.contents = value
            self.available = True
            self.condition.notify_all()  # Avisa al consumidor

File: Lab01/ejercicios_resueltos/Anexo2/Java/CubbyHole.java
package Java;

public class CubbyHole {
    private int contents;
    private boolean available = false;

    // El Consumidor llama a este método
    public synchronized int get() {
        while (available == false) {
            try {
                wait(); // Espera si el cajón está vacío
            } catch (InterruptedException e) { }
        }
        available = false;
        notifyAll(); // Despierta al productor
        return contents;
    }

    // El Productor llama a este método
    public synchronized void put(int value) {
        while (available == true) {
            try {
                wait(); // Espera si el cajón ya tiene algo
            } catch (InterruptedException e) { }
        }
        contents = value;
        available = true;
        notifyAll(); // Despierta al consumidor
    }
}